### ISSUE TITLE
Inevitable unhandled Promise rejection due to failure of prefetch

### ISSUE THREAD

**Issue by uhyo** [webpack/webpack#14115]
# Bug report

TL; DR: [Promise without `.catch` here.](https://github.com/webpack/webpack/blob/179d2e00fe0b901ff176f1c6dbbf6a17f1ff44ea/lib/prefetch/ChunkPrefetchTriggerRuntimeModule.js#L38)

**What is the current behavior?**

If you dynamic-import a chunk and that chunk tries to prefetch another chunk, and that prefetch fails for some reason, you get an unhandled Promise rejection reported even if you are doing proper error handling.

**If the current behavior is a bug, please provide the steps to reproduce.**

Here is a reproduction repo: https://github.com/uhyo/webpack-prefetch-rejection-repro

1. Build by `npx webpack`.
2. Open `index.html`.
3. You see an unhandled Promise rejection reported.
<img width="300" alt="スクリーンショット 2021-09-02 23 10 22" src="https://user-images.githubusercontent.com/748348/131859055-9f5f3dca-0aba-4213-a9a9-f732b62da1b7.png">

The repro includes a custom Webpack plugin (`BoomPlugin`) that simulates a failure of loading of accompanying resources. Of course, this is just for easily reproducing the issue; the case I actually faced involved mini-css-extract-plugin trying to load a CSS chunk accompanying main chunk.

What is of note is that all `import()` calls included in source code properly handles Promise rejections. The above image shows how a prefetch failure error is handled by `import(...).catch` but is still reported as a Promise rejection via another  path.

Seems that this unhandled Promise rejection is inevitable; the rejecting Promise is created internally in Webpack runtime and there is no way of preventing that.

**What is the expected behavior?**

Properly handle the error internally, or provide a way to register a `.catch` handler to the internal Promise.

**Other relevant information:**
webpack version:  5.51.1
Node.js version: 14.15.0
Operating System: MacoS Mojave 10.14.6
Additional tools:

**Comment by alexander-akait** [comment]
You can solve this on your application side

**Comment by uhyo** [comment]
@alexander-akait I think I did the best in the application code in the repro. All application-side Promises have proper error handling.

Do you mean that you don't think this is not a problem of Webpack runtime? Then I will bring a bug report to mini-css-webpack-plugin side.

**Comment by uhyo** [comment]
Hey, do I have a chance of second opinion?
I will wait for another week and then close this issue as not a bug and bring a issue to mini-css-extract-plugin that chunk loading should never fail on their side.

**Comment by alexander-akait** [comment]
> Properly handle the error internally, or provide a way to register a .catch handler to the internal Promise.

You already doing this using:
```
import("./split.js").catch((err) => {
  console.error("import failed", err)
})
```

I don't see any problems with logic here, if you have `unhandled Promise rejection` with `mini-css-extract-plugin` please open an issue in `mini-css-extract-plugin` repo and provide example, I agree you should have ability to `catch`

**Comment by alexander-akait** [comment]
oh, I see, do you want to catch...

**Comment by alexander-akait** [comment]
Interesting case, can you provide more information why you need to catch this? If something was not loaded it can break the application logic, so we prefer to throw an error always in this case

**Comment by alexander-akait** [comment]
Maybe related https://github.com/webpack-contrib/mini-css-extract-plugin/issues/802

**Comment by uhyo** [comment]
Thank you for noticing :smiley:

I prepared another example that is less self-contained but is closer to actual situation. Sorry, I should have presented this from first.

https://github.com/uhyo/webpack-prefetch-rejection-repro-2

1. Build by `npx webpack`.
2. Open `index.html`.
3. Use devtools to block request to `src_split2_js.css`.

<img width="778" alt="スクリーンショット 2021-09-12 14 07 24" src="https://user-images.githubusercontent.com/748348/132972962-f4ead3f0-6250-4c73-837d-db4c6b691505.png">

4. Then reopen `index.html` and check console.
<img width="778" alt="スクリーンショット 2021-09-12 14 08 23" src="https://user-images.githubusercontent.com/748348/132972971-cf958953-4606-499f-b6ff-8b625e74f807.png">

5. Despite all the effort to catch errors, Uncaught error is printed to console.

So when `css` files emitted by `mini-css-extract-plugin` failed to load, the Promise returned by `import()` rejects. This is totally fine, but there is another path where rejecting Promise is passed from `mini-css-extract-plugin` runtime to webpack runtime leading to unhandled Promise rejection, which application code has no chance of intercepting.

The actual problem we're facing is that this unhandled rejection error is recorded in our error logging facility and we can't suppress the error (expect for letting the error happen but not sending it).

Error being thrown is fine, but inevitable Promise rejection is not great I think :cry:

**Comment by alexander-akait** [comment]
I will look at this in near future, sorry for delay

**Comment by alexander-akait** [comment]
@uhyo Interesting case, we decided - if something cannot be loaded, it should interrupt the application (related, but not the same https://github.com/webpack-contrib/mini-css-extract-plugin/issues/802). Just interesting - how it happens?

In theory you can use dirty workaround, put:
```
let original = __webpack_require__.f.prefetch;

__webpack_require__.f.prefetch = (...args) => {
  original(args).catch(() => {}); 
};
```

in `index.js`. 

I don't have strong opinion here, on the one hand you are right - you should have ability to catch, on the other hand why, it was failed, don't think you can load it again. But there is case when you can load fallback. Can you provide information why you need this, what is the real case?

**Comment by uhyo** [comment]
@alexander-akait thank you for your investigation!

> Can you provide information why you need this, what is the real case?

Our use case is error logging. Failure of CSS chunk loading may happen, and we properly handle the error (we give up loading that resource and show a nice error page to the user). However, our error logging facility (namely [Sentry](https://sentry.io/welcome/)) still logs it as “uncaught runtime error” because of the unhandled Promise rejection reported here, which still happens even though we `.catch()` it in our application code.

We could set up the error logging facility so that we ignore “loading CSS chunk failed” errors, but we consider it as another dirty workaround. Recently unhandled Promise rejections are considered a sort of uncaught runtime errors; for example, Node.js changed the default setting to terminating the process as soon as an unhandled Promise rejection happens. Therefore, we want to remove the root cause of unhandled Promise rejections whenever possible.

I hope this answers your question 🥺

**Comment by alexander-akait** [comment]
Sounds reasonable, marked as bug, feel free to send a fix

**Comment by alexander-akait** [comment]
bump

**Comment by mnpenner** [comment]
I've got the exact same use-case as @uhyo. These errors are finding their way into my Sentry logs and spamming me.

Not really sure what causes them because all the ones I've seen, I try going to the URL and it does exist. Perhaps some race condition when I deploy a new version of the app or their browser is cancelling the request?

Regardless, a way to catch them so I can handle them better would be nice.

FWIW, here's what I see:

![image](https://user-images.githubusercontent.com/1112998/160048876-1735880c-627f-48e4-9d5b-5dd8bf1b4b7e.png)

**Comment by alexander-akait** [comment]
@vankop Can you look at this?

**Comment by vankop** [comment]
So.. if you try to do "inner" prefetch ( e.g. in case entrypoint => `import(a)` => `import(/* webpackPrefetch: true */b)`) you will get `Uncaught (in promise) ChunkLoadError` when you executing entrypoint and this is quite valid since chunk `b` was failed to prefetch and no `.catch` for it was executed yet..

When code execution will reach `import(/* webpackPrefetch: true */b).catch(e => {})` catch will be executed with `ChunkLoadingError` (thats expected).

I understand that you want to deduplicate errors in Sentry (or other observability tool), but there is scenario when prefetch fails, but chunk never was requested (no `import(b)` was executed)

**Comment by uhyo** [comment]
@vankop So I understand that prefetch failure cannot always be tied to ` import ()` calls, but closing this issue without any solution isn't quite right. 😢

Currently there is no way of handling failure of prefetch which leads to inevitable error logs on Sentry. How about adding an official way of handling prefetch errors?

**Comment by vankop** [comment]
> Currently there is no way of handling failure of prefetch which leads to inevitable error logs on Sentry.

you can use `.catch` still, it will catch error **only** when execution will reach `import()`

**Comment by alexander-akait** [comment]
@vankop `catch` is not help here, because we don't rethrow an exception, try to repo above

**Comment by vankop** [comment]
@alexander-akait catch works file
<img width="1266" alt="Screenshot 2022-03-28 at 13 24 23" src="https://user-images.githubusercontent.com/10380560/160378866-2b00531e-d903-4a5f-a7c4-d4cb8a950a9f.png">

index.js
```js
console.log("Hello, world!")

import("./split.js").catch((err) => {
  console.log("import failed")
})
```

split.js
```js
console.log("This is split chunk");

import(/* webpackPrefetch: true */"./split2.js").catch((error) => {
  console.log("import2 failed!");
})
```

split2.js
```js
console.log("prefetched chunk")
```

after compile just remove `src_split2_js.js` chunk and run `index.html`

**Comment by vankop** [comment]
right now there is a bug that `split.js` also prefetch need a review fix https://github.com/webpack/webpack/pull/15393
